test/src/basic/array_helper.h

```diff
@@ -35,189 +35,503 @@
+				for (iterator it = array->begin(); it != array->end(); it++) {
+					AssertThat(counter, IsLessThan(list.size()));
+					AssertThat((*array)[it.key()], Equals(maybeWrap<ElementType>(counter)));
+					AssertThat(*it, Equals(maybeWrap<ElementType>(counter)));
+					*it = maybeWrap<ElementType>(counter + list.size());
+					if (maybeWrap<ElementType>(counter + list.size()) == fillElement) {
+						defaults++;
+					}
+					counter++;
+				}
+				AssertThat(counter, Equals(list.size()));
+
+				counter = 0;
+				for (const_iterator it = cArray.cbegin(); it != cArray.cend(); it++) {
+					AssertThat(counter, IsLessThan(list.size()));
+					AssertThat((*array)[it.key()],
+							Equals(maybeWrap<ElementType>(counter + list.size())));
+					counter++;
+				}
+				AssertThat(counter, Equals(list.size()));
+
+				for (KeyType val : list) {
+					if ((*array)[val] == fillElement) {
+						defaults--;
+					}
+				}
+				AssertThat(defaults, Equals(0));
+			});
+		});
+	});
+}
+
+/**
+ * Perform basic tests for a map of registered keys to non copy constructible values.
+ * @tparam BaseType the type of the class (e.g. Graph) that maintains the registered keys
+ * @tparam ArrayType the type of array to be tested
+ * @tparam KeyType the type of registered key
+ * @tparam ElementType the value type
+ *
+ * @param title the title of the top-level bandit::describe
+ * @param initBase a function to initialize the base
+ * @param chooseKey a function to choose an arbitrary key from the base
+ * @param getAllKeys a function to generate a list of all keys
+ * @param createKey a function to create a new key in the base
+ */
+template<class BaseType, template<typename, bool> class ArrayType, typename KeyType, typename ElementType>
+void describeArrayWithoutDefault(const std::string& title, std::function<void(BaseType&)> initBase,
+		std::function<KeyType(const BaseType&)> chooseKey,
+		std::function<void(const BaseType&, List<KeyType>&)> getAllKeys,
+		std::function<KeyType(BaseType&)> createKey) {
+	using MyArrayType = ArrayType<ElementType, false>;
+	using RegistryType = const typename MyArrayType::registry_type;
+	using const_iterator = typename MyArrayType::const_iterator;
+	using iterator = typename MyArrayType::iterator;
+
+	describe(title.c_str(), [&]() {
+		std::unique_ptr<MyArrayType> array;
+		BaseType base;
+		initBase(base);
+
+		before_each([&]() {
+			array.reset(new MyArrayType());
+			initBase(base);
+		});
+
+		it("handles nested arrays well", [&]() {
+			BaseType B;
+			initBase(B);
+
+			List<KeyType> keys;
+			getAllKeys(B, keys);
+
+			ArrayType<MyArrayType, false> nestedArray(B);
+			for (KeyType k : keys) {
+				nestedArray[k].init(B);
+			}
+		});
+
+		describe("init", [&]() {
+			it("initializes w/o a registry", [&]() {
+				AssertThat(array->registeredAt(), IsNull());
+				AssertThat(array->valid(), IsFalse());
+				array->init();
+				AssertThat(array->registeredAt(), IsNull());
+				AssertThat(array->valid(), IsFalse());
+			});
```

Comment:
I tried doing this with templates, but there are some slight differences with how default values are created and handled, which made my two attempts at doing this fail. It might now be easier with C++17 `constexpr if`, but I settled for being ok with this slight duplication (although Sonar also still complains about this).